File: sdk/typescript/src/forge-provider.ts
import type { Extension } from "./extension.js";
import type { ExtensionContext } from "./extension-contract.js";
import { InvalidParamsError } from "./errors.js";
import { registerProvideSurface } from "./extension-provide-surface.js";
import type {
  ForgeCapabilitiesRequest,
  ForgeCapabilitiesResponse,
  ForgePRCreateRequest,
  ForgePRCreateResponse,
  ForgeStatusRequest,
  ForgeStatusResponse,
} from "./types.js";

export const FORGE_PROVIDER_CAPABILITY = "forge.provider";
export const FORGE_CAPABILITIES_METHOD = "forge/capabilities";
export const FORGE_STATUS_METHOD = "forge/status";
export const FORGE_PR_CREATE_METHOD = "forge/pr_create";

export interface ForgeProviderHandlers {
  capabilities: (
    context: ExtensionContext,
    request: ForgeCapabilitiesRequest
  ) => Promise<ForgeCapabilitiesResponse> | ForgeCapabilitiesResponse;
  status: (
    context: ExtensionContext,
    request: ForgeStatusRequest
  ) => Promise<ForgeStatusResponse> | ForgeStatusResponse;
  createPR: (
    context: ExtensionContext,
    request: ForgePRCreateRequest
  ) => Promise<ForgePRCreateResponse> | ForgePRCreateResponse;
}

function requestRecord(method: string, request: unknown): Record<string, unknown> {
  if (typeof request !== "object" || request === null || Array.isArray(request)) {
    throw new InvalidParamsError(`${method} params must be an object`);
  }
  return request as Record<string, unknown>;
}

function requiredString(method: string, request: Record<string, unknown>, field: string): string {
  const value = request[field];
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new InvalidParamsError(`${method} requires a non-empty ${field}`);
  }
  return value;
}

function remoteURLs(method: string, request: Record<string, unknown>): string[] {
  const values = request.remote_urls;
  if (
    !Array.isArray(values) ||
    values.length === 0 ||
    values.some(value => typeof value !== "string" || value.trim().length === 0)
  ) {
    throw new InvalidParamsError(`${method} requires non-empty remote_urls`);
  }
  return values;
}

function capabilitiesRequest(request: unknown): ForgeCapabilitiesRequest {
  const record = requestRecord(FORGE_CAPABILITIES_METHOD, request);
  return { remote_urls: remoteURLs(FORGE_CAPABILITIES_METHOD, record) };
}

function statusRequest(request: unknown): ForgeStatusRequest {
  const record = requestRecord(FORGE_STATUS_METHOD, request);
  return {
    remote_urls: remoteURLs(FORGE_STATUS_METHOD, record),
    branch: requiredString(FORGE_STATUS_METHOD, record, "branch"),
  };
}

function createPRRequest(request: unknown): ForgePRCreateRequest {
  const record = requestRecord(FORGE_PR_CREATE_METHOD, request);
  const parsed: ForgePRCreateRequest = {
    remote_urls: remoteURLs(FORGE_PR_CREATE_METHOD, record),
    head: requiredString(FORGE_PR_CREATE_METHOD, record, "head"),
    base: requiredString(FORGE_PR_CREATE_METHOD, record, "base"),
    title: requiredString(FORGE_PR_CREATE_METHOD, record, "title"),
  };
  if (record.body !== undefined) {
    if (typeof record.body !== "string") {
      throw new InvalidParamsError(`${FORGE_PR_CREATE_METHOD} body must be a string`);
    }
    parsed.body = record.body;
  }
  if (record.draft !== undefined) {
    if (typeof record.draft !== "boolean") {
      throw new InvalidParamsError(`${FORGE_PR_CREATE_METHOD} draft must be a boolean`);
    }
    parsed.draft = record.draft;
  }
  return parsed;
}

export function registerForgeProvider(
  extension: Extension,
  handlers: ForgeProviderHandlers
): Extension {
  if (!extension) {
    throw new Error("extension is required");
  }
  if (
    typeof handlers?.capabilities !== "function" ||
    typeof handlers.status !== "function" ||
    typeof handlers.createPR !== "function"
  ) {
    throw new Error("forge provider handlers are required");
  }
  return extension[registerProvideSurface](FORGE_PROVIDER_CAPABILITY, [
    {
      method: FORGE_CAPABILITIES_METHOD,
      handler: async (context, request) =>
        await handlers.capabilities(context, capabilitiesRequest(request)),
    },
    {
      method: FORGE_STATUS_METHOD,
      handler: async (context, request) => await handlers.status(context, statusRequest(request)),
    },
    {
      method: FORGE_PR_CREATE_METHOD,
      handler: async (context, request) =>
        await handlers.createPR(context, createPRRequest(request)),
    },
  ]);
}

File: sdk/go/forge_provider.go
package compozysdk

import (
	"context"
	"encoding/json"
	"strings"

	"github.com/compozy/compozy/sdk/go/contracts"
)

// ForgeCapabilitiesRequest asks which remotes a provider serves.
type ForgeCapabilitiesRequest = contracts.ForgeCapabilitiesRequest

// ForgeCapabilitiesResponse describes provider vocabulary and availability.
type ForgeCapabilitiesResponse = contracts.ForgeCapabilitiesResponse

// ForgeStatusRequest asks for pull-request state by branch.
type ForgeStatusRequest = contracts.ForgeStatusRequest

// ForgeStatusResponse reports the current pull-request state.
type ForgeStatusResponse = contracts.ForgeStatusResponse

// ForgePRCreateRequest asks a provider to create or find one pull request.
type ForgePRCreateRequest = contracts.ForgePRCreateRequest

// ForgePRCreateResponse reports the created or existing pull request.
type ForgePRCreateResponse = contracts.ForgePRCreateResponse

// ForgeProviderHandlers implements the daemon-initiated forge provider services.
type ForgeProviderHandlers struct {
	Capabilities func(context.Context, ExtensionContext, ForgeCapabilitiesRequest) (ForgeCapabilitiesResponse, error)
	Status       func(context.Context, ExtensionContext, ForgeStatusRequest) (ForgeStatusResponse, error)
	CreatePR     func(context.Context, ExtensionContext, ForgePRCreateRequest) (ForgePRCreateResponse, error)
}

// ForgeProvider registers one public forge provider atomically.
func ForgeProvider(extension *Extension, handlers ForgeProviderHandlers) error {
	if extension == nil {
		return NewInternalError("extension is required")
	}
	if handlers.Capabilities == nil || handlers.Status == nil || handlers.CreatePR == nil {
		return NewInvalidParamsError("forge provider handlers are required", nil)
	}
	registrations := []struct {
		method  string
		handler ExtensionHandler
	}{
		{ExtensionServiceMethodForgeCapabilities, typedForgeHandler(handlers.Capabilities)},
		{ExtensionServiceMethodForgeStatus, typedForgeHandler(handlers.Status)},
		{ExtensionServiceMethodForgePRCreate, typedForgeHandler(handlers.CreatePR)},
	}
	extension.mu.Lock()
	defer extension.mu.Unlock()
	if extension.initialized {
		return NewInvalidParamsError("extension registration is closed after initialize", nil)
	}
	if extension.forgeProvider {
		return NewInvalidParamsError("forge provider is already registered", nil)
	}
	for _, registration := range registrations {
		if _, exists := extension.handlers[registration.method]; exists {
			return NewInvalidParamsError(registration.method+" is reserved by ForgeProvider", nil)
		}
	}
	for _, registration := range registrations {
		extension.handlers[registration.method] = registration.handler
		extension.transport.Handle(registration.method, extension.dispatch)
	}
	extension.forgeProvider = true
	extension.definition.Capabilities.Provides = normalizeStrings(append(
		extension.definition.Capabilities.Provides,
		CapabilityProvideForgeProvider,
	))
	return nil
}

func typedForgeHandler[Request any, Response any](
	handler func(context.Context, ExtensionContext, Request) (Response, error),
) ExtensionHandler {
	return func(ctx context.Context, extensionContext ExtensionContext, raw json.RawMessage) (any, error) {
		var request Request
		if len(raw) == 0 {
			return nil, NewInvalidParamsError("forge request is required", nil)
		}
		if err := json.Unmarshal(raw, &request); err != nil {
			return nil, NewInvalidParamsError("forge request is invalid", map[string]any{"error": err.Error()})
		}
		if err := validateForgeProviderRequest(any(request)); err != nil {
			return nil, err
		}
		return handler(ctx, extensionContext, request)
	}
}

func validateForgeProviderRequest(request any) error {
	requireRemotes := func(remotes []string) error {
		if len(remotes) == 0 {
			return NewInvalidParamsError("forge remote_urls are required", nil)
		}
		for _, remote := range remotes {
			if strings.TrimSpace(remote) == "" {
				return NewInvalidParamsError("forge remote_urls cannot contain empty values", nil)
			}
		}
		return nil
	}
	switch value := request.(type) {
	case ForgeCapabilitiesRequest:
		return requireRemotes(value.RemoteURLs)
	case ForgeStatusRequest:
		if err := requireRemotes(value.RemoteURLs); err != nil {
			return err
		}
		if strings.TrimSpace(value.Branch) == "" {
			return NewInvalidParamsError("forge branch is required", nil)
		}
	case ForgePRCreateRequest:
		if err := requireRemotes(value.RemoteURLs); err != nil {
			return err
		}
		if strings.TrimSpace(value.Head) == "" || strings.TrimSpace(value.Base) == "" ||
			strings.TrimSpace(value.Title) == "" {
			return NewInvalidParamsError("forge head, base, and title are required", nil)
		}
	}
	return nil
}

func isForgeProviderMethod(method string) bool {
	switch strings.TrimSpace(method) {
	case ExtensionServiceMethodForgeCapabilities,
		ExtensionServiceMethodForgeStatus,
		ExtensionServiceMethodForgePRCreate:
		return true
	default:
		return false
	}
}
